### PR TITLE
[5981] Showing users avatars while browsing charts.

### PR DESCRIPTION
**Pull request by tomaszszymanski129**
Added showing user faces, modified users API, created script for updating local XML file (bin/update-xml) created on buildout.
```
Name                      Stmts   Miss  Cover   Missing
-------------------------------------------------------
presence_analyzer             4      0   100%   
presence_analyzer.main        5      0   100%   
presence_analyzer.urls        8      0   100%   
presence_analyzer.utils      49      3    94%   60, 67-68
presence_analyzer.views      51      0   100%   
-------------------------------------------------------
TOTAL                       117      3    98%   
----------------------------------------------------------------------
Ran 12 tests in 0.193s
```
https://bugzilla.stxnext.pl/show_bug.cgi?id=5981

### CODE REVIEW

File: src/presence_analyzer/utils.py [hunk 1]
```diff
@@ -35,23 +36,27 @@ def inner(*args, **kwargs):
...
-                'start': datetime.time(8, 30, 0),
-                'end': datetime.time(16, 45, 0),
-            },
-        }
-    }
+    Contains name and avatar if available in XML file.
     """
     data = {}
     with open(app.config['DATA_CSV'], 'r') as csvfile:
+        try:
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: Is there a need to edit ```get_data```? As of now it returns strictly what is needed to the presence views charts, there is a separate view for the dropdown. Why not make a separate function for it?

File: src/presence_analyzer/views.py [hunk 2]
```diff
@@ -40,10 +40,26 @@ def users_view():
...
     """
     data = get_data()
-    return [
-        {'user_id': i, 'name': 'User {0}'.format(str(i))}
-        for i in data.keys()
-    ]
+    result = []
+    for key, value in data.iteritems():
+        name = value['credentials']['name']
+        if name:
```
Comment (mjelonek92) [inline src/presence_analyzer/views.py]: This XML contains the only users that we want to display. CSV for the dropdown is not needed anymore. Thus every users should have a name.

File: src/presence_analyzer/utils.py [hunk 3]
```diff
@@ -35,23 +36,27 @@ def inner(*args, **kwargs):
...
+            xml_file = open(app.config['DATA_XML'], 'r')
+            tree = etree.parse(xml_file)
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: ```parse``` also accepts file path. If you want to use file object then you need to open it in binary mode first.

File: src/presence_analyzer/templates/presence_start_end.html [hunk 4]
```diff
@@ -27,19 +26,28 @@
...
                 });
                 $('#user_id').change(function() {
+                    document.getElementById('avatar').innerHTML = "";
                     var selected_user = $("#user_id").val(),
                         chart_div = $('#chart_div');
                     if (selected_user) {
+                        var current = avatars[selected_user];
+                        if (current) {
+                            var img = document.createElement("img");
+                            img.src = "https://intranet.stxnext.pl" + current;
```
Comment (mjelonek92) [inline src/presence_analyzer/templates/presence_start_end.html]: Please do not hard code the address. Have the backend send the whole url. The XML contains the server address.

File: src/presence_analyzer/update_xml.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    log = logging.getLogger(__name__)
+    url = urllib2.urlopen(app.config['XML_URL'])
+    tmp = 'users.xml.tmp'
+
+    with open(tmp, 'wb') as output:
+        try:
+            output.write(url.read())
+            os.rename(tmp, app.config['DATA_XML'])
+        except OSError:
+                log.error('OS error occurred.')
```
Comment (mjelonek92) [inline src/presence_analyzer/update_xml.py]: Wrong indentation.

File: src/presence_analyzer/update_xml.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    """
+    Downloads new XML database, then moves it to proper place.
+    """
+    etc = partial(os.path.join, 'parts', 'etc')
+    cfg = etc('deploy.cfg')
+
+    app.config.from_pyfile(os.path.abspath(cfg))
+    log = logging.getLogger(__name__)
+    url = urllib2.urlopen(app.config['XML_URL'])
+    tmp = 'users.xml.tmp'
```
Comment (mjelonek92) [inline src/presence_analyzer/update_xml.py]: Can you not write directly to users.xml?

File: src/presence_analyzer/update_xml.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+from functools import partial
+
+from presence_analyzer import app
+
+
+def update():
+    """
+    Downloads new XML database, then moves it to proper place.
+    """
+    etc = partial(os.path.join, 'parts', 'etc')
```
Comment (mjelonek92) [inline src/presence_analyzer/update_xml.py]: Why the need for partial?

**Comment by dpolcyn** [comment]
Please add text, if we don't have data for selected user - "No data for this user"

File: src/presence_analyzer/update_xml.py (hunk 6 above)
Comment (tomaszszymanski129) [inline src/presence_analyzer/update_xml.py]: And how am I writing directly to users.xml here? Its a temporary file, downloaded first, moved to destination afterwards.

Comment (mjelonek92) [inline src/presence_analyzer/update_xml.py]: That's what I'm asking. Why do you need the temporary file and not just write directly to the one you need?

File: src/presence_analyzer/templates/mean_time_weekday.html [hunk 8]
```diff
@@ -27,39 +16,55 @@
...
-                            chart_div.show();
-                            loading.hide();
-                            var chart = new google.visualization.ColumnChart(chart_div[0]);
-                            chart.draw(data, options);
+                                chart_div.show();
+                                loading.hide();
+                                var chart = new google.visualization.ColumnChart(chart_div[0]);
+                                chart.draw(data, options);
+                            }
+                            else {
```
Comment (dpolcyn) [inline src/presence_analyzer/templates/mean_time_weekday.html]: ```} else {```

File: src/presence_analyzer/tests.py [hunk 9]
```diff
@@ -65,8 +69,17 @@ def test_api_users(self):
...
         self.assertEqual(resp.content_type, 'application/json')
         data = json.loads(resp.data)
-        self.assertEqual(len(data), 2)
-        self.assertDictEqual(data[0], {'user_id': 10, 'name': 'User 10'})
+        self.assertEqual(len(data), 10)
+        self.assertDictEqual(
+            data[0], {
+                'avatar': '/api/images/users/36',
+                'user_id': 36,
+                'name': 'Anna W.'}
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: Put that closing brace on the next line.

File: src/presence_analyzer/views.py [hunk 10]
```diff
@@ -39,10 +48,28 @@ def users_view():
...
+    except IOError:
+        log.warning(
+            'No user data XML file found. '
+            'You can download it by running \"bin/update_xml\"'
+        )
+        xml_users = {}
     return [
-        {'user_id': i, 'name': 'User {0}'.format(str(i))}
-        for i in data.keys()
+        {
```
Comment (mjelonek92) [inline src/presence_analyzer/views.py]: Can you not do this at the stage where you're defining ```xml_users```? It looks like repetitive work and you don't use ```xml_users``` anywhere else.

File: src/presence_analyzer/views.py [hunk 11]
```diff
@@ -39,10 +48,28 @@ def users_view():
...
+        tree = etree.parse(app.config['DATA_XML'])
+        users_from_xml = tree.getroot().find('users')
+        xml_users = {
+            int(user.get('id')): {
+                'name': user.find('name').text,
+                'avatar': user.find('avatar').text
+            } for user in users_from_xml
+        }
+    except IOError:
+        log.warning(
```
Comment (mjelonek92) [inline src/presence_analyzer/views.py]: This looks to me as more like an error/exception than a warning.

**Pull request merged**